File: sentry-android-core/src/main/java/io/sentry/android/core/internal/gestures/WindowCallbackAdapter.java
package io.sentry.android.core.internal.gestures;

import android.annotation.SuppressLint;
import android.view.ActionMode;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MenuItem;
import android.view.MotionEvent;
import android.view.SearchEvent;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.view.accessibility.AccessibilityEvent;
import com.jakewharton.nopen.annotation.Open;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

@Open
public class WindowCallbackAdapter implements Window.Callback {

  private final @NotNull Window.Callback delegate;

  public WindowCallbackAdapter(final Window.@NotNull Callback delegate) {
    this.delegate = delegate;
  }

  @Override
  public boolean dispatchKeyEvent(KeyEvent keyEvent) {
    return delegate.dispatchKeyEvent(keyEvent);
  }

  @Override
  public boolean dispatchKeyShortcutEvent(KeyEvent keyEvent) {
    return delegate.dispatchKeyShortcutEvent(keyEvent);
  }

  @Override
  public boolean dispatchTouchEvent(@Nullable MotionEvent motionEvent) {
    return delegate.dispatchTouchEvent(motionEvent);
  }

  @Override
  public boolean dispatchTrackballEvent(MotionEvent motionEvent) {
    return delegate.dispatchTrackballEvent(motionEvent);
  }

  @Override
  public boolean dispatchGenericMotionEvent(MotionEvent motionEvent) {
    return delegate.dispatchGenericMotionEvent(motionEvent);
  }

  @Override
  public boolean dispatchPopulateAccessibilityEvent(AccessibilityEvent accessibilityEvent) {
    return delegate.dispatchPopulateAccessibilityEvent(accessibilityEvent);
  }

  @Nullable
  @Override
  public View onCreatePanelView(int i) {
    return delegate.onCreatePanelView(i);
  }

  @Override
  public boolean onCreatePanelMenu(int i, @NotNull Menu menu) {
    return delegate.onCreatePanelMenu(i, menu);
  }

  @Override
  public boolean onPreparePanel(int i, @Nullable View view, @NotNull Menu menu) {
    return delegate.onPreparePanel(i, view, menu);
  }

  @Override
  public boolean onMenuOpened(int i, @NotNull Menu menu) {
    return delegate.onMenuOpened(i, menu);
  }

  @Override
  public boolean onMenuItemSelected(int i, @NotNull MenuItem menuItem) {
    return delegate.onMenuItemSelected(i, menuItem);
  }

  @Override
  public void onWindowAttributesChanged(WindowManager.LayoutParams layoutParams) {
    delegate.onWindowAttributesChanged(layoutParams);
  }

  @Override
  public void onContentChanged() {
    delegate.onContentChanged();
  }

  @Override
  public void onWindowFocusChanged(boolean b) {
    delegate.onWindowFocusChanged(b);
  }

  @Override
  public void onAttachedToWindow() {
    delegate.onAttachedToWindow();
  }

  @Override
  public void onDetachedFromWindow() {
    delegate.onDetachedFromWindow();
  }

  @Override
  public void onPanelClosed(int i, @NotNull Menu menu) {
    delegate.onPanelClosed(i, menu);
  }

  @Override
  public boolean onSearchRequested() {
    return delegate.onSearchRequested();
  }

  @SuppressLint("NewApi")
  @Override
  public boolean onSearchRequested(SearchEvent searchEvent) {
    return delegate.onSearchRequested(searchEvent);
  }

  @Nullable
  @Override
  public ActionMode onWindowStartingActionMode(ActionMode.Callback callback) {
    return delegate.onWindowStartingActionMode(callback);
  }

  @SuppressLint("NewApi")
  @Nullable
  @Override
  public ActionMode onWindowStartingActionMode(ActionMode.Callback callback, int i) {
    return delegate.onWindowStartingActionMode(callback, i);
  }

  @Override
  public void onActionModeStarted(ActionMode actionMode) {
    delegate.onActionModeStarted(actionMode);
  }

  @Override
  public void onActionModeFinished(ActionMode actionMode) {
    delegate.onActionModeFinished(actionMode);
  }
}

File: sentry-android-core/src/main/java/io/sentry/android/core/replay/WindowCallbackAdapter.kt
package io.sentry.android.core.replay

import android.view.ActionMode
import android.view.KeyEvent
import android.view.Menu
import android.view.MenuItem
import android.view.MotionEvent
import android.view.SearchEvent
import android.view.View
import android.view.Window
import android.view.WindowManager
import android.view.accessibility.AccessibilityEvent

open class WindowCallbackDelegate(callback: Window.Callback?) : Window.Callback {

    private val original: Window.Callback

    init {
        original = callback ?: EmptyCallback()
    }

    override fun dispatchKeyEvent(event: KeyEvent?): Boolean {
        return original.dispatchKeyEvent(event)
    }

    override fun dispatchKeyShortcutEvent(event: KeyEvent?): Boolean {
        return original.dispatchKeyShortcutEvent(event)
    }

    override fun dispatchTouchEvent(event: MotionEvent?): Boolean {
        return original.dispatchTouchEvent(event)
    }

    override fun dispatchTrackballEvent(event: MotionEvent?): Boolean {
        return original.dispatchTrackballEvent(event)
    }

    override fun dispatchGenericMotionEvent(event: MotionEvent?): Boolean {
        return original.dispatchGenericMotionEvent(event)
    }

    override fun dispatchPopulateAccessibilityEvent(event: AccessibilityEvent?): Boolean {
        return original.dispatchPopulateAccessibilityEvent(event)
    }

    override fun onCreatePanelView(featureId: Int): View? {
        return original.onCreatePanelView(featureId)
    }

    override fun onCreatePanelMenu(featureId: Int, menu: Menu): Boolean {
        return original.onCreatePanelMenu(featureId, menu)
    }

    override fun onPreparePanel(featureId: Int, view: View?, menu: Menu): Boolean {
        return original.onPreparePanel(featureId, view, menu)
    }

    override fun onMenuOpened(featureId: Int, menu: Menu): Boolean {
        return original.onMenuOpened(featureId, menu)
    }

    override fun onMenuItemSelected(featureId: Int, item: MenuItem): Boolean {
        return original.onMenuItemSelected(featureId, item)
    }

    override fun onWindowAttributesChanged(attrs: WindowManager.LayoutParams?) {
        return original.onWindowAttributesChanged(attrs)
    }

    override fun onContentChanged() {
        return original.onContentChanged()
    }

    override fun onWindowFocusChanged(hasFocus: Boolean) {
        return original.onWindowFocusChanged(hasFocus)
    }

    override fun onAttachedToWindow() {
        return original.onAttachedToWindow()
    }

    override fun onDetachedFromWindow() {
        return original.onDetachedFromWindow()
    }

    override fun onPanelClosed(featureId: Int, menu: Menu) {
        return original.onPanelClosed(featureId, menu)
    }

    override fun onSearchRequested(): Boolean {
        return original.onSearchRequested()
    }

    override fun onSearchRequested(searchEvent: SearchEvent?): Boolean {
        return original.onSearchRequested(searchEvent)
    }

    override fun onWindowStartingActionMode(callback: ActionMode.Callback?): ActionMode? {
        return original.onWindowStartingActionMode(callback)
    }

    override fun onWindowStartingActionMode(
        callback: ActionMode.Callback?,
        type: Int
    ): ActionMode? {
        return original.onWindowStartingActionMode(callback, type)
    }

    override fun onActionModeStarted(mode: ActionMode?) {
        return original.onActionModeStarted(mode)
    }

    override fun onActionModeFinished(mode: ActionMode?) {
        return original.onActionModeFinished(mode)
    }
}

class EmptyCallback : Window.Callback {
    override fun dispatchKeyEvent(event: KeyEvent?): Boolean {
        return false
    }

    override fun dispatchKeyShortcutEvent(event: KeyEvent?): Boolean = false

    override fun dispatchTouchEvent(event: MotionEvent?): Boolean = false

    override fun dispatchTrackballEvent(event: MotionEvent?): Boolean = false

    override fun dispatchGenericMotionEvent(event: MotionEvent?): Boolean = false

    override fun dispatchPopulateAccessibilityEvent(event: AccessibilityEvent?): Boolean = false

    override fun onCreatePanelView(featureId: Int): View? = null

    override fun onCreatePanelMenu(featureId: Int, menu: Menu): Boolean = false

    override fun onPreparePanel(featureId: Int, view: View?, menu: Menu): Boolean = false

    override fun onMenuOpened(featureId: Int, menu: Menu): Boolean = false

    override fun onMenuItemSelected(featureId: Int, item: MenuItem): Boolean = false

    override fun onWindowAttributesChanged(attrs: WindowManager.LayoutParams?) {
    }

    override fun onContentChanged() {
    }

    override fun onWindowFocusChanged(hasFocus: Boolean) {
    }

    override fun onAttachedToWindow() {
    }

    override fun onDetachedFromWindow() {
    }

    override fun onPanelClosed(featureId: Int, menu: Menu) {
    }

    override fun onSearchRequested(): Boolean = false

    override fun onSearchRequested(searchEvent: SearchEvent?): Boolean = false

    override fun onWindowStartingActionMode(callback: ActionMode.Callback?): ActionMode? = null

    override fun onWindowStartingActionMode(
        callback: ActionMode.Callback?,
        type: Int
    ): ActionMode? = null

    override fun onActionModeStarted(mode: ActionMode?) {
    }

    override fun onActionModeFinished(mode: ActionMode?) {
    }
}
